README.md
# NAME

Dancer2::Plugin::OpenAPI - A Dancer2 plugin for creating routes from an OpenAPI spec

# VERSION

version 0.003\_900

# SYNOPSIS

`example/my_app.pl`:

    #!/usr/bin/env perl

    use Dancer2;
    use Dancer2::Plugin::OpenAPI;
    use Path::Tiny;

    openapi( url => path(__FILE__)->sibling('openapi.yaml') );

    sub my_controller {
        return "Hello World!\n";
    }

    dance;

`example/openapi.yaml`:

    ---
    swagger: "2.0"
    info:
      title: MyApp's API
      version: "1.0"
    basePath: /api
    paths:
      /welcome:
        get:
          operationId: my_controller
          responses:
            200:
              description: success

Then on the terminal run:

    perl my_app.pl
    curl http://localhost:3000/api/welcome

You'll find the example files displayed above in the distribution and repository.

# MIGRATING FROM DANCER1

If you've been using Dancer1 before you might know [Dancer::Plugin::Swagger](https://metacpan.org/pod/Dancer::Plugin::Swagger).
Please note that that module's workflow is completely different! It is about
creating the spec from the app. The module described in this text is about
reading the spec and creating parts of the app for you.

# DEBUGGING

To see some more debug messages on STDERR set environment variable `OPENAPI_DEBUG`
to a true value.

# METHODS

## openapi( url => $url, ... )

Import routes from OpenAPI file. Named arguments:

- `url`: URL to passed to [Swagger2](https://metacpan.org/pod/Swagger2) module
- `controller_factory`: custom callback generator/finder that returns callbacks to routes
- `create_options_route`: autocreate additional route replying to OPTIONS requests based on Swagger data
- `validate_spec`: boolish value (default: true) telling if Swagger2 file shall be validated by official Swagger specification
- `validate_requests`: boolish value (default: same as `validate_spec`) telling if HTTP requests shall be validated by loaded specification (needs `validate_spec` to be true)
- `validate_responses`: boolish value (default: same as `validate_spec`) telling if HTTP responses shall be validated by loaded specification (needs `validate_spec` to be true)

## default\_controller\_factory

Default method for finding a callback for a given `operationId`. Can be
overriden by the `controller_factory` argument to `openapi()` or config option.

The default uses the `controller` argument/config option or the name of
the app (possibly with `::Controller` appended). If the `operationId`
looks like a Perl module the module name is tried under the namespace
mentioned above and as a top level module name.

The module warns as long as controller modules or methods can't be found
and returns a coderef on the first match.

# SEE ALSO

- [Mojolicious::Plugin::OpenAPI](https://metacpan.org/pod/Mojolicious::Plugin::OpenAPI) A similar plugin for the [Mojolicious](https://metacpan.org/pod/Mojolicious) Web framework
- [http://swagger.io/](http://swagger.io/): Website of the Swagger alias OpenAPI Specification

# ACKNOWLEDGEMENT

This software has been developed with support from [STRATO](https://www.strato.com/).
In German: Diese Software wurde mit Unterstützung von [STRATO](https://www.strato.de/) entwickelt.

# AUTHORS

- Daniel Böhmer <dboehmer@cpan.org>
- Tina Müller <cpan2@tinita.de>

# COPYRIGHT AND LICENSE

This software is copyright (c) 2016 by Daniel Böhmer.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
Revision history for Perl module Dancer2::Plugin::OpenAPI

TO BE RELEASED

  * rename from Dancer2::Plugin::Swagger2 to Dancer2::Plugin::OpenAPI
  * move from Swagger2 to JSON::Validator::OpenAPI

TO BE RELEASED

  * Added feature to create a HTTP OPTIONS route.

0.03 2016-04-07

  * Allowed request/response validation with invalid spec.
  * Changed used Dist::Zilla plugins.
  * Began encouraging usage of Path::Tiny.
  * Extended documentation.

0.02 2016-03-16

  * Improved distribution metadata, e.g. requirements.
  * Fixed filename in documentation.

0.01 2016-03-10

  * First release to CPAN.


lib/Dancer2/Plugin/Swagger2.pm
package Dancer2::Plugin::Swagger2;

use strict;
use warnings;

# ABSTRACT: A Dancer2 plugin for creating routes from a Swagger2 spec
# VERSION

use Dancer2::Plugin;
use Module::Runtime 'use_module';
use Swagger2;
use Swagger2::SchemaValidator;

=head1 MIGRATING FROM DANCER1

If you've been using Dancer1 before you might know L<Dancer::Plugin::Swagger>.
Please note that that module's workflow is completely different! It is about
creating the spec from the app. The module described in this text is about
reading the spec and creating parts of the app for you.

=head1 SYNOPSIS

C<example/my_app.pl>:

{{ include('example/my_app.pl')->indent }}

C<example/swagger2.yaml>:

{{ include('example/swagger2.yaml')->indent }}

Then on the terminal run:

    perl my_app.pl
    curl http://localhost:3000/api/welcome

You'll find the example files displayed above in the distribution and repository.

=head1 DEBUGGING

To see some more debug messages on STDERR set environment variable C<SWAGGER2_DEBUG>
to a true value.

=cut

sub DEBUG { !!$ENV{SWAGGER2_DEBUG} }

# for definition see http://www.json.org/number.gif
sub RE_JSON_NUMBER { qr/ ^ -? \d+ ( \. \d+ )? ( [eE] [+-]? \d+ )? $ /x }

=head1 METHODS

=head2 swagger2( url => $url, ... )

Import routes from Swagger file. Named arguments:

=over

=item * C<url>: URL to passed to L<Swagger2> module

=item * C<controller_factory>: custom callback generator/finder that returns callbacks to routes

=item * C<create_options_route>: autocreate additional route replying to OPTIONS requests based on Swagger data

=item * C<validate_spec>: boolish value (default: true) telling if Swagger2 file shall be validated by official Swagger specification

=item * C<validate_requests>: boolish value (default: same as C<validate_spec>) telling if HTTP requests shall be validated by loaded specification (needs C<validate_spec> to be true)

=item * C<validate_responses>: boolish value (default: same as C<validate_spec>) telling if HTTP responses shall be validated by loaded specification (needs C<validate_spec> to be true)

=back

=cut

register swagger2 => sub {
    my ( $dsl, %args ) = @_;
    my $conf = plugin_setting;

    ### get arguments/config values/defaults ###

    my $controller_factory =
         $args{controller_factory} || \&_default_controller_factory;
    my $url = $args{url} or die "argument 'url' missing";
    my $create_options_route =
        exists $args{create_options_route}   ? !!$args{create_options_route}
      : exists $conf->{create_options_route} ? !!$conf->{create_options_route}
      :                                        '';
    my $validate_spec =
        exists $args{validate_spec}   ? !!$args{validate_spec}
      : exists $conf->{validate_spec} ? !!$conf->{validate_spec}
      :                                 1;
    my $validate_requests =
        exists $args{validate_requests}   ? !!$args{validate_requests}
      : exists $conf->{validate_requests} ? !!$conf->{validate_requests}
      :                                     $validate_spec;
    my $validate_responses =
        exists $args{validate_responses}   ? !!$args{validate_responses}
      : exists $conf->{validate_responses} ? !!$conf->{validate_responses}
      :                                      $validate_spec;

    # parse Swagger2 file
    my $spec = Swagger2->new($url)->expand;

    if ( $validate_spec or $validate_requests or $validate_responses ) {
        if ( my @errors = $spec->validate ) {
            if ($validate_spec) {
                die join "\n" => "Swagger2: Invalid spec:", @errors;
            }
            else {
                warn "Spec contains errors but"
                  . " request/response validation is enabled!";
            }
        }
    }

    my $basePath = $spec->api_spec->get('/basePath');
    my $paths    = $spec->api_spec->get('/paths');    # TODO might be undef?

    while ( my ( $path => $path_spec ) = each %$paths ) {
        my $dancer2_path = $path;

        $basePath and $dancer2_path = $basePath . $dancer2_path;

        # adapt Swagger2 syntax for URL path arguments to Dancer2 syntax
        # '/path/{argument}' -> '/path/:argument'
        $dancer2_path =~ s/\{([^{}]+?)\}/:$1/g;

        my @http_methods = sort keys %$path_spec;

        if ($create_options_route) {
            my $allow_methods = join ', ' => 'OPTIONS', map uc, @http_methods;
            $dsl->options(
                $dancer2_path => sub {
                    $dsl->headers(
                        Allow => $allow_methods,    # RFC 2616 HTTP/1.1
                        'Access-Control-Allow-Methods' => $allow_methods, # CORS
                        'Access-Control-Max-Age'       => 60 * 60 * 24,
                    );
                },
            );
        }

        for my $http_method (@http_methods) {
            my $method_spec = $path_spec->{ $http_method };
            my $coderef = $controller_factory->(
                $method_spec, $http_method, $path, $dsl, $conf, \%args
            ) or next;

            DEBUG and warn "Add route $http_method $dancer2_path";

            my $params = $method_spec->{parameters};

            # Dancer2 DSL keyword is different from HTTP method
            $http_method eq 'delete' and $http_method = 'del';

            $dsl->$http_method(
                $dancer2_path => sub {
                    my @args = @_;

                    if ($validate_requests) {
                        my @errors =
                          _validate_request( $method_spec, $dsl->app->request );

                        if (@errors) {
                            DEBUG and warn "Invalid request: @errors\n";
                            $dsl->status(400);
                            return { errors => [ map { "$_" } @errors ] };
                        }
                    }

                    my $result = $coderef->(@args);

                    if ($validate_responses) {
                        my @errors =
                          _validate_response( $method_spec, $dsl->response,
                            $result );

                        if (@errors) {
                            DEBUG and warn "Invalid response: @errors\n";
                            $dsl->status(500);

                            # TODO hide details of server-side errors?
                            return { errors => [ map { "$_" } @errors ] };
                        }
                    }

                    return $result;
                }
            );
        }
    }
};

register_plugin;

sub _validate_request {
    my ( $method_spec, $request ) = @_;

    my @errors;

    for my $parameter_spec ( @{ $method_spec->{parameters} } ) {
        my $in       = $parameter_spec->{in};
        my $name     = $parameter_spec->{name};
        my $required = $parameter_spec->{required};

        if ( $in eq 'body' ) {    # complex data structure in HTTP body
            my $input  = $request->data;
            my $schema = $parameter_spec->{schema};

            push @errors, _validator()->validate_input( $input, $schema );
        }
        else {    # simple key-value-pair in HTTP header/query/path/form
            my $type = $parameter_spec->{type};
            my @values;

            if ( $in eq 'header' ) {
                @values = $request->header($name);
            }
            elsif ( $in eq 'query' ) {
                @values = $request->query_parameters->get_all($name);
            }
            elsif ( $in eq 'path' ) {
                @values = $request->route_parameters->get_all($name);
            }
            elsif ( $in eq 'formData' ) {
                @values = $request->body_parameters->get_all($name);
            }
            else { die "Unknown value for property 'in' of parameter '$name'" }

            # TODO align error messages to output style of SchemaValidator
            if ( @values == 0 and $required ) {
                $required and push @errors, "No value for parameter '$name'";
                next;
            }
            elsif ( @values > 1 ) {
                push @errors, "Multiple values for parameter '$name'";
                next;
            }

            my $value = $values[0] // $parameter_spec->{default};

            # TODO steal more from Mojolicious::Plugin::Swagger2 ;-)
            if ( ( $type eq 'integer' or $type eq 'number' )
                and $value =~ RE_JSON_NUMBER )
            {
                $value += 0;    # numify
            }
            elsif ( $type eq 'boolean' ) {
                $value = ( !$value or $value eq 'false' ) ? '' : 1;
            }

            my %input  = defined $value ? ( $name => $value ) : ();
            my %schema = ( properties => { $name => $parameter_spec } );

            $required and $schema{required} = [$name];

            push @errors, _validator()->validate_input( \%input, \%schema );
        }
    }

    return @errors;
}

sub _validate_response {
    my ( $method_spec, $response, $result ) = @_;

    my $responses = $method_spec->{responses};
    my $status    = $response->status;

    my @errors;

    if ( my $response_spec = $responses->{$status} || $responses->{default} ) {

        my $headers = $response_spec->{headers};

        while ( my ( $name => $header_spec ) = each %$headers ) {
            my @values = $response->header($name);

            if ( $header_spec->{type} eq 'array' ) {
                push @errors,
                  _validator()->validate_input( \@values, $header_spec );
            }
            else {
                if ( @values == 0 ) {
                    next;    # you can't make a header 'required' in Swagger2
                }
                elsif ( @values > 1 ) {

                   # TODO align error message to output style of SchemaValidator
                    push @errors, "header '$name' has multiple values";
                    next;
                }

                push @errors,
                  _validator()->validate_input( $values[0], $header_spec );
            }
        }

        if ( my $schema = $response_spec->{schema} ) {
            push @errors, _validator()->validate_input( $result, $schema );
        }
    }
    else {
        # TODO Call validate_input($response, {}) like
        #      in Mojolicious::Plugin::Swagger2?
        # Swagger2-0.71/lib/Mojolicious/Plugin/Swagger2.pm line L315
    }

    return @errors;
}

=head2 default_controller_factory

Default method for finding a callback for a given C<operationId>. Can be
overriden by the C<controller_factory> argument to C<swagger2> or config option.

The default uses the C<controller> argument/config option or the name of
the app (possibly with C<::Controller> appended). If the C<operationId>
looks like a Perl module the module name is tried under the namespace
mentioned above and as a top level module name.

The module warns as long as controller modules or methods can't be found
and returns a coderef on the first match.

=cut

sub _default_controller_factory {
    # TODO simplify argument list
    my ( $method_spec, $http_method, $path, $dsl, $conf, $args, ) = @_;

    # from Dancer2 app
    my $namespace = $args->{controller} || $conf->{controller};
    my $app = $dsl->app->name;

    # from Swagger2 file
    my $module;
    my $method = $method_spec->{operationId};
    if ( $method =~ s/^(.+)::// ) {    # looks like Perl module
        $module = $1;
    }

    # different candidates possibly reflecting operationId
    my @controller_candidates = do {
        if ($namespace) {
            if ($module) { $namespace . '::' . $module, $module }
            else         { $namespace }
        }
        else {
            if ($module) {
                (                      # parens for better layout by Perl::Tidy
                    $app . '::' . $module,
                    $app . '::Controller::' . $module,
                    $module,           # maybe a top level module name?
                );
            }
            else { $app, $app . '::Controller' }
        }
    };

    # check candidates
    for my $controller (@controller_candidates) {
        local $@;
        if ( ! eval { use_module( $controller ); 1; } ) {
            if ( $@ && $@ =~ m/^Can't locate / ) {    # module doesn't exist
                DEBUG and warn "Can't load '$controller'";

                # don't do `next` here because controller could be
                # defined in other package ...
            }
            else {    # module doesn't compile
                die $@;
            }
        }

        if ( my $cb = $controller->can($method) ) {
            return $cb;    # confirmed candidate
        }
        else {
            DEBUG and warn "Controller '$controller' can't '$method'";
        }
    }

    # none found
    warn "Can't find any handler for operationId '$method_spec->{operationId}'";
    return;
}

my $validator;
sub _validator { $validator ||= Swagger2::SchemaValidator->new }

=head1 SEE ALSO

=over

=item * L<Mojolicious::Plugin::Swagger2> A similar plugin for the L<Mojolicious> Web framework

=item * L<http://swagger.io/>: Website of the Swagger alias OpenAPI Specification

=back

=head1 ACKNOWLEDGEMENT

This software has been developed with support from L<STRATO|https://www.strato.com/>.
In German: Diese Software wurde mit Unterstützung von L<STRATO|https://www.strato.de/> entwickelt.

=cut

1;


lib/Dancer2/Plugin/OpenAPI.pm
package Dancer2::Plugin::OpenAPI;

use strict;
use warnings;

# ABSTRACT: A Dancer2 plugin for creating routes from an OpenAPI spec
# VERSION

use Dancer2::Plugin;
use JSON::Validator;
use JSON::Validator::OpenAPI::Dancer2;
use Module::Runtime 'use_module';

=head1 MIGRATING FROM DANCER1

If you've been using Dancer1 before you might know L<Dancer::Plugin::Swagger>.
Please note that that module's workflow is completely different! It is about
creating the spec from the app. The module described in this text is about
reading the spec and creating parts of the app for you.

=head1 SYNOPSIS

C<example/my_app.pl>:

{{ include('example/my_app.pl')->indent }}

C<example/openapi.yaml>:

{{ include('example/openapi.yaml')->indent }}

Then on the terminal run:

    perl my_app.pl
    curl http://localhost:3000/api/welcome

You'll find the example files displayed above in the distribution and repository.

=head1 DEBUGGING

To see some more debug messages on STDERR set environment variable C<OPENAPI_DEBUG>
to a true value.

=cut

sub DEBUG { !!$ENV{OPENAPI_DEBUG} }

=head1 METHODS

=head2 openapi( url => $url, ... )

Import routes from OpenAPI file. Named arguments:

=over

=item * C<url>: URL to passed to L<Swagger2> module

=item * C<controller_factory>: custom callback generator/finder that returns callbacks to routes

=item * C<create_options_route>: autocreate additional route replying to OPTIONS requests based on Swagger data

=item * C<validate_spec>: boolish value (default: true) telling if Swagger2 file shall be validated by official Swagger specification

=item * C<validate_requests>: boolish value (default: same as C<validate_spec>) telling if HTTP requests shall be validated by loaded specification (needs C<validate_spec> to be true)

=item * C<validate_responses>: boolish value (default: same as C<validate_spec>) telling if HTTP responses shall be validated by loaded specification (needs C<validate_spec> to be true)

=back

=cut

register openapi => sub {
    my ( $dsl, %args ) = @_;
    my $conf = plugin_setting;

    ### get arguments/config values/defaults ###

    my $controller_factory =
      $args{controller_factory} || \&_default_controller_factory;
    my $url = $args{url} or die "argument 'url' missing";
    my $create_options_route =
        exists $args{create_options_route}   ? !!$args{create_options_route}
      : exists $conf->{create_options_route} ? !!$conf->{create_options_route}
      :                                        '';
    my $validate_requests =
        exists $args{validate_requests}   ? !!$args{validate_requests}
      : exists $conf->{validate_requests} ? !!$conf->{validate_requests}
      :                                     1;
    my $validate_responses =
        exists $args{validate_responses}   ? !!$args{validate_responses}
      : exists $conf->{validate_responses} ? !!$conf->{validate_responses}
      :                                      1;

    # depcrecation notice
    if ( exists $args{validate_spec} or exists $conf->{validate_spec} ) {
        warn "Config key 'validate_spec' is not supported anymore";
    }

    # validate $url against OpenAPI spec
    # TODO is this step done implicitly in the next step?
    my $openapi =
      JSON::Validator->new->schema(JSON::Validator::OpenAPI::SPECIFICATION_URL);
    $openapi->validate($url);

    my $api = JSON::Validator->new->schema($url);

    my $basePath = $api->schema->get('/basePath');
    my $paths    = $api->schema->get('/paths');      # TODO might be undef?

    while ( my ( $path => $path_spec ) = each %$paths ) {
        my $dancer2_path = $path;

        $basePath and $dancer2_path = $basePath . $dancer2_path;

        # adapt Swagger2 syntax for URL path arguments to Dancer2 syntax
        # '/path/{argument}' -> '/path/:argument'
        $dancer2_path =~ s/\{([^{}]+?)\}/:$1/g;

        my @http_methods = sort keys %$path_spec;

        if ($create_options_route) {
            my $allow_methods = join ', ' => 'OPTIONS', map uc, @http_methods;
            $dsl->options(
                $dancer2_path => sub {
                    $dsl->headers(
                        Allow                          => $allow_methods,    # RFC 2616 HTTP/1.1
                        'Access-Control-Allow-Methods' => $allow_methods,    # CORS
                        'Access-Control-Max-Age'       => 60 * 60 * 24,
                    );
                },
            );
        }

        for my $http_method (@http_methods) {
            my $method_spec = $path_spec->{$http_method};
            my $coderef     = $controller_factory->(
                $method_spec, $http_method, $path, $dsl, $conf, \%args
            ) or next;

            DEBUG and warn "Add route $http_method $dancer2_path";

            my $params = $method_spec->{parameters};

            # Dancer2 DSL keyword is different from HTTP method
            $http_method eq 'delete' and $http_method = 'del';

            $dsl->$http_method(
                $dancer2_path => sub {
                    my @args = @_;

                    if ($validate_requests) {
                        my @errors = _validator()->validate_request( $dsl, $method_spec );

                        if (@errors) {
                            DEBUG and warn "Invalid request: @errors\n";
                            $dsl->status(400);
                            return { errors => [ map { "$_" } @errors ] };
                        }
                    }

                    my $result = $coderef->(@args);

                    if ($validate_responses) {
                        my @errors = _validator()
                          ->validate_response( $dsl, $method_spec, $dsl->response->status, $result );

                        if (@errors) {
                            DEBUG and warn "Invalid response: @errors\n";
                            $dsl->status(500);

                            # TODO hide details of server-side errors?
                            return { errors => [ map { "$_" } @errors ] };
                        }
                    }

                    return $result;
                }
            );
        }
    }
};

register_plugin;

=head2 default_controller_factory

Default method for finding a callback for a given C<operationId>. Can be
overriden by the C<controller_factory> argument to C<openapi()> or config option.

The default uses the C<controller> argument/config option or the name of
the app (possibly with C<::Controller> appended). If the C<operationId>
looks like a Perl module the module name is tried under the namespace
mentioned above and as a top level module name.

The module warns as long as controller modules or methods can't be found
and returns a coderef on the first match.

=cut

sub _default_controller_factory {

    # TODO simplify argument list
    my ( $method_spec, $http_method, $path, $dsl, $conf, $args, ) = @_;

    # from Dancer2 app
    my $namespace = $args->{controller} || $conf->{controller};
    my $app = $dsl->app->name;

    # from Swagger2 file
    my $module;
    my $method = $method_spec->{operationId};
    if ( $method =~ s/^(.+)::// ) {    # looks like Perl module
        $module = $1;
    }

    # different candidates possibly reflecting operationId
    my @controller_candidates = do {
        if ($namespace) {
            if ($module) { $namespace . '::' . $module, $module }
            else         { $namespace }
        }
        else {
            if ($module) {
                (                      # parens for better layout by Perl::Tidy
                    $app . '::' . $module,
                    $app . '::Controller::' . $module,
                    $module,           # maybe a top level module name?
                );
            }
            else { $app, $app . '::Controller' }
        }
    };

    # check candidates
    for my $controller (@controller_candidates) {
        local $@;
        if ( !eval { use_module($controller); 1; } ) {
            if ( $@ && $@ =~ m/^Can't locate / ) {    # module doesn't exist
                DEBUG and warn "Can't load '$controller'";

                # don't do `next` here because controller could be
                # defined in other package ...
            }
            else {    # module doesn't compile
                die $@;
            }
        }

        if ( my $cb = $controller->can($method) ) {
            return $cb;    # confirmed candidate
        }
        else {
            DEBUG and warn "Controller '$controller' can't '$method'";
        }
    }

    # none found
    warn "Can't find any handler for operationId '$method_spec->{operationId}'";
    return;
}

my $validator;
sub _validator { $validator ||= JSON::Validator::OpenAPI::Dancer2->new }

=head1 SEE ALSO

=over

=item * L<Mojolicious::Plugin::OpenAPI> A similar plugin for the L<Mojolicious> Web framework

=item * L<http://swagger.io/>: Website of the Swagger alias OpenAPI Specification

=back

=head1 ACKNOWLEDGEMENT

This software has been developed with support from L<STRATO|https://www.strato.com/>.
In German: Diese Software wurde mit Unterstützung von L<STRATO|https://www.strato.de/> entwickelt.

=cut

1;
